tests/Formatters/CharsetFormatterTests.cs

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Tools.Formatters;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Tools.Tests.Formatters
+{
+    public class CharsetFormatterTests : CSharpFormatterTests
+    {
+        private protected override ICodeFormatter Formatter => new CharsetFormatter();
+
+        [Theory]
+        [InlineData("latin1", "utf-8")]
+        [InlineData("latin1", "utf-8-bom")]
+        [InlineData("latin1", "utf-16be")]
+        [InlineData("latin1", "utf-16le")]
+        [InlineData("utf-8", "latin1")]
+        [InlineData("utf-8", "utf-8-bom")]
+        [InlineData("utf-8", "utf-16be")]
+        [InlineData("utf-8", "utf-16le")]
+        [InlineData("utf-8-bom", "latin1")]
+        [InlineData("utf-8-bom", "utf-8")]
+        [InlineData("utf-8-bom", "utf-16be")]
+        [InlineData("utf-8-bom", "utf-16le")]
+        [InlineData("utf-16be", "latin1")]
+        [InlineData("utf-16be", "utf-8")]
+        [InlineData("utf-16be", "utf-8-bom")]
+        [InlineData("utf-16be", "utf-16le")]
+        [InlineData("utf-16le", "latin1")]
+        [InlineData("utf-16le", "utf-8")]
+        [InlineData("utf-16le", "utf-8-bom")]
+        [InlineData("utf-16le", "utf-16be")]
+        public async Task TestCharsetWrong_CharsetFixed(string codeValue, string expectedValue)
+        {
+            var codeEncoding = CharsetFormatter.GetCharset(codeValue);
+            var expectedEncoding = CharsetFormatter.GetCharset(expectedValue);
+
+            var testCode = "class C { }";
```

Comment:
It might be worthwhile to verify that the encoding really changed the test code contents, perhaps by making the test class name a more complex character, and including a byte[] with the final raw content in the InlineData

Reply:
Yeah, we will need to make an integration test for that. There wouldn't be any change until the workspace has saved the document changes. Since these tests use AdhocWorkspace that isn't possible because it doesn't implement persistence.